# Package

In [40]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from ISLP import load_data
from ISLP.models import ModelSpec as MS
from ISLP import confusion_table
from sklearn.preprocessing import StandardScaler
import pandas as pd
from sklearn.model_selection import train_test_split
import numpy as np

# Dataset

In [19]:
Smarket = load_data('Smarket')
Smarket.drop("Direction", axis=1).corr()
Smarket.head()

,Year,Lag1,Lag2,Lag3,Lag4,Lag5,Volume,Today,Direction
0,2001,0.381,-0.192,-2.624,-1.055,5.010,1.1913,0.959,Up
1,2001,0.959,0.381,-0.192,-2.624,-1.055,1.2965,1.032,Up
2,2001,1.032,0.959,0.381,-0.192,-2.624,1.4112,-0.623,Down
3,2001,-0.623,1.032,0.959,0.381,-0.192,1.2760,0.614,Up
4,2001,0.614,-0.623,1.032,0.959,0.381,1.2057,0.213,Up


In [20]:
# Dataframe
allvars = Smarket.columns[1:3]
design = MS(allvars)
X = design.fit_transform(Smarket)
X.head()

,intercept,Lag1,Lag2
0,1.0,0.381,-0.192
1,1.0,0.959,0.381
2,1.0,1.032,0.959
3,1.0,-0.623,1.032
4,1.0,0.614,-0.623


In [21]:
# Train and Test
train = (Smarket.Year < 2005)
X_train, X_test = X.loc[train], X.loc[~train]

In [22]:
# Cible
D = Smarket.Direction
L_train, L_test = D.loc[train], D.loc[~train]

# Exercice 1

In [23]:
# instancier la classe de KNN 
knn1 = KNeighborsClassifier(n_neighbors=1)

# Entrainer le KNN
knn1.fit(X_train, L_train)

# Prédiction de nouvelle donnée X_test avec KNN 
knn_pred = knn1.predict(X_test)

# Evaluation du KNN 
confusion_table(knn_pred, L_test)

Truth,Down,Up
Predicted,,
Down,43,58
Up,68,83


In [24]:
print("La bonne prédiction est de :", ((43 + 83) / (43 + 68 + 58 +83)) * 100, "%")

La bonne prédiction est de : 50.0 %


Ce n'est pas une bonne prédiction. 

In [25]:
# KNN avec K = 3
## Instancier KNN
knn3 = KNeighborsClassifier(n_neighbors = 3)

## Entrainement du KNN
knn3.fit(X_train, L_train)

## Prédiction avec le KNN 
knn3_pred = knn3.predict(X_test)

## Evaluation
CT = confusion_table(knn3_pred, L_test)

In [26]:
CT

Truth,Down,Up
Predicted,,
Down,48,55
Up,63,86


In [27]:
correct_predictions = CT.at["Down", "Down"] + CT.at["Up", "Up"]
total_predictions = CT.values.sum()
accuracy = (correct_predictions / total_predictions) * 100
accuracy

np.float64(53.17460317460318)

# Exercice 2

In [28]:
Caravan = load_data("Caravan")
print(Caravan.shape)
print(Caravan.head())

(5822, 86)
   MOSTYPE  MAANTHUI  MGEMOMV  MGEMLEEF  MOSHOOFD  MGODRK  MGODPR  MGODOV  \
0       33         1        3         2         8       0       5       1   
1       37         1        2         2         8       1       4       1   
2       37         1        2         2         8       0       4       2   
3        9         1        3         3         3       2       3       2   
4       40         1        4         2        10       1       4       1   

   MGODGE  MRELGE  ...  APERSONG  AGEZONG  AWAOREG  ABRAND  AZEILPL  APLEZIER  \
0       3       7  ...         0        0        0       1        0         0   
1       4       6  ...         0        0        0       1        0         0   
2       4       3  ...         0        0        0       1        0         0   
3       4       5  ...         0        0        0       1        0         0   
4       4       7  ...         0        0        0       1        0         0   

   AFIETS  AINBOED  ABYSTAND  Purchase 

In [29]:
Caravan = load_data('Caravan')
Purchase = Caravan.Purchase
Purchase.value_counts()

Purchase
No     5474
Yes     348
Name: count, dtype: int64

In [30]:
feature_df = Caravan.drop(columns="Purchase")

In [31]:
# Standartiser
scaler = StandardScaler(with_mean = True, # Moyenne égale à 0
                        with_std = True, #standard deviation égale à 1
                        copy = True) #utilisation du df

In [32]:
scaler.fit(feature_df)
X_std = scaler.transform(feature_df)

In [33]:
feature_std = pd.DataFrame(
    X_std, 
    columns= feature_df.columns
)

feature_std.describe()

,MOSTYPE,MAANTHUI,MGEMOMV,MGEMLEEF,MOSHOOFD,MGODRK,MGODPR,MGODOV,MGODGE,MRELGE,...,ALEVEN,APERSONG,AGEZONG,AWAOREG,ABRAND,AZEILPL,APLEZIER,AFIETS,AINBOED,ABYSTAND
count,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,...,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5.822000e+03,5822.000000,5.822000e+03,5.822000e+03
mean,-7.749822e-17,-1.614038e-16,-2.099164e-16,2.154084e-16,-2.929067e-17,-1.220444e-17,7.932889e-17,-1.403511e-17,9.519466e-17,-1.476738e-16,...,-4.515644e-17,2.379867e-17,1.464533e-17,9.763555e-18,-1.183831e-16,-6.102222e-19,-1.830667e-18,0.000000,3.905422e-17,3.356222e-18
std,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,...,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086e+00,1.000086,1.000086e+00,1.000086e+00
min,-1.810219e+00,-2.725800e-01,-2.125697e+00,-2.444683e+00,-1.671134e+00,-6.943106e-01,-2.696827e+00,-1.051594e+00,-2.039739e+00,-3.238560e+00,...,-2.029099e-01,-7.316512e-02,-8.105460e-02,-5.992001e-02,-1.014358e+00,-2.270578e-02,-7.365027e-02,-0.150620,-8.734772e-02,-1.188165e-01
25%,-1.109590e+00,-2.725800e-01,-8.595001e-01,-1.216964e+00,-9.709796e-01,-6.943106e-01,-3.654101e-01,-1.051594e+00,-7.877902e-01,-6.198244e-01,...,-2.029099e-01,-7.316512e-02,-8.105460e-02,-5.992001e-02,-1.014358e+00,-2.270578e-02,-7.365027e-02,-0.150620,-8.734772e-02,-1.188165e-01
50%,4.473633e-01,-2.725800e-01,4.066966e-01,1.075466e-02,4.293284e-01,-6.943106e-01,2.174440e-01,-6.871064e-02,-1.618158e-01,-9.607727e-02,...,-2.029099e-01,-7.316512e-02,-8.105460e-02,-5.992001e-02,7.649707e-01,-2.270578e-02,-7.365027e-02,-0.150620,-8.734772e-02,-1.188165e-01
75%,8.366015e-01,-2.725800e-01,4.066966e-01,1.075466e-02,7.794054e-01,3.025516e-01,8.002982e-01,9.141723e-01,4.641586e-01,4.276698e-01,...,-2.029099e-01,-7.316512e-02,-8.105460e-02,-5.992001e-02,7.649707e-01,-2.270578e-02,-7.365027e-02,-0.150620,-8.734772e-02,-1.188165e-01
max,1.303687e+00,2.190544e+01,2.939090e+00,3.693911e+00,1.479559e+00,8.277450e+00,2.548861e+00,3.862821e+00,3.594031e+00,1.475164e+00,...,2.098707e+01,1.366772e+01,1.233736e+01,2.578114e+01,1.144094e+01,4.404165e+01,2.442874e+01,14.069579,2.202302e+01,1.654985e+01


In [34]:
# Split
(X_train,
 X_test,
 y_train,
 y_test) = train_test_split(feature_std, 
                            Purchase,
                            test_size=1000,
                            random_state=0)

In [35]:
knn1 = KNeighborsClassifier(n_neighbors=1)
knn1.fit(X_train, y_train)
knn1_pred = knn1.predict(X_test)
# np.mean converti le booléen de True en 1 et False en 0
np.mean(y_test != knn1_pred), np.mean(y_test != "No")

(np.float64(0.111), np.float64(0.067))

Le taux d'erreur est dans 1000 observations est de 11%. 
Le modèle ne performe pas puisque le taux d'achat est estimé à 6,7%. 

çà veut dire que si on essaye de vendre l'assurance aléatoirement au client, seul 6% qui achètent le produit. C'est pas intéressant. 

Le plus important est d'identifier correctement les acheteurs. 

In [37]:
confusion_table(knn1_pred, y_test)

Truth,No,Yes
Predicted,,
No,880,58
Yes,53,9


Dans 62 clients (53 + 9), ou 14,6% vont acheter l'assurance. C'est le double de ce qu'on obtient à partir d'un random model (6,7%). 

# Tuning Parameters

In [39]:
for K in range(1,6):
    knn = KNeighborsClassifier(n_neighbors=K)
    knn_pred = knn.fit(X_train, y_train).predict(X_test)
    C = confusion_table(knn_pred, y_test)
    templ = ('K={0:d}: # predicted to rent: {1:>2},' + ' # who did rent {2:d}, accuracy {3:.1%}')
    pred = C.loc['Yes'].sum()
    did_rent = C.loc['Yes', 'Yes']
    print(templ.format(
        K,
        pred,
        did_rent,
        did_rent / pred
    ))

K=1: # predicted to rent: 62, # who did rent 9, accuracy 14.5%
K=2: # predicted to rent:  6, # who did rent 1, accuracy 16.7%
K=3: # predicted to rent: 20, # who did rent 3, accuracy 15.0%
K=4: # predicted to rent:  4, # who did rent 0, accuracy 0.0%
K=5: # predicted to rent:  7, # who did rent 1, accuracy 14.3%


# Comparison to Logistic Regression

In [41]:
logit = LogisticRegression(C = 1e10, solver = 'liblinear')
logit.fit(X_train, y_train)
logit_pred = logit.predict_proba(X_test)
logit_labels = np.where(logit_pred[:,1] > 5, 'Yes', 'No')
confusion_table(logit_labels, y_test)

Truth,No,Yes
Predicted,,
No,933,67
Yes,0,0


In [42]:
logit_labels = np.where(logit_pred[:, 1] > 0.25, 'Yes', 'No')
confusion_table(logit_labels, y_test)

Truth,No,Yes
Predicted,,
No,913,58
Yes,20,9


In [43]:
9 / (20 + 9)

0.3103448275862069

31% d'inscription dans l'assurance a été prédite par le modèle. 